components/mpas-ocean/src/shared/mpas_ocn_vertical_remap.F

```diff
@@ -0,0 +1,466 @@
+      do iEdge = 1, nEdgesAll
+         kmin = minLevelEdgeBot(iEdge)
+         kmax = maxLevelEdgeTop(iEdge)
+         cell1 = cellsOnEdge(1,iEdge)
+         cell2 = cellsOnEdge(2,iEdge)
+         do k = 1,nVertLevels
+            ! initialize layerThicknessEdgeMean to avoid divide by
+            ! zero and NaN problems.
+            layerThickEdgeNew(k,iEdge) = -1.0e34_RKIND
+         end do
+         do k = kmin,kmax
+            ! central differenced
+            layerThickEdgeNew(k,iEdge) = 0.5_RKIND * &
+                                         (layerThicknessNew(k,cell1) + &
+                                          layerThicknessNew(k,cell2))
+         end do
+      end do
+#ifndef MPAS_OPENACC
+      !$omp end do
+      !$omp end parallel
+#endif
+
+      !$omp parallel
+      !$omp do schedule(runtime) private(k,kTop,kBot,cell1,cell2)
+      do iEdge = 1, nEdges
+
+         cell1 = cellsOnEdge(1,iEdge)
+         cell2 = cellsOnEdge(2,iEdge)
+
+         kTop = minLevelEdgeBot(iEdge)
+         kBot = maxLevelEdgeTop(iEdge)
+
+         ! We only need to define the kBot+1 depth for one active cell
+         ! as the absolute kBot+1 depth has no effect on remapping
+         if (cell1 <= nCellsAll .and. maxLevelCell(cell1) > 0) then
+            heightEdgeNow(:, iEdge) = heightCellNow(kBot+1, cell1)
+            heightEdgeNew(:, iEdge) = heightCellNew(kBot+1, cell1)
+         elseif (cell2 <= nCellsAll .and. maxLevelCell(cell2) > 0) then
+            heightEdgeNow(:, iEdge) = heightCellNow(kBot+1, cell2)
+            heightEdgeNew(:, iEdge) = heightCellNew(kBot+1, cell2)
+         else
+            call mpas_log_write('No valid cell for edge remapping', MPAS_LOG_WARN)
+            heightEdgeNow(:,iEdge) = 0.0_RKIND
+            cycle
+         endif
+
+         do k = kBot, kTop, -1
+            heightEdgeNow(k, iEdge) = heightEdgeNow(k+1, iEdge) - layerThickEdgeMean(k, iEdge)
+            heightEdgeNew(k, iEdge) = heightEdgeNew(k+1, iEdge) - layerThickEdgeNew(k, iEdge)
+         end do
+
+      end do
+      !$omp end do
+      !$omp end parallel
+
+      ! Assign new layerThicknesses
+      layerThickness = layerThicknessNew
+
+      !----------------------------------------------------------------------------
+      ! ACTUAL REMAPPING FROM HERE
+
+      nDoFs = 1 ! always 1 for MPAS, it'd be >1 for DG methods (with more DoF per layer)
+
+      ! how many variables to be remapped concurrently per column
+      if (config_use_freq_filtered_thickness) then
+         nVars = 3
+      else
+         nVars = 1
+      endif
+
+      allocate(dataNow(nDoFs, nVars, nVertLevels+1))
+      allocate(dataNew(nDoFs, nVars, nVertLevels+1))
+      allocate(bcUpper(nVars))
+      allocate(bcLower(nVars))
+
+      ! set boundary conditions
+      bcUpper%bcopt = bc_upper
+      bcLower%bcopt = bc_lower
+
+      dataNow = 0.0_RKIND
+      dataNew = 0.0_RKIND
+
+      call work%init(nVertLevels + 1, nVars, opts) ! this is internal workspace for PPR
```

Comment:
@cbegeman @mark-petersen this call to `work%init` is allocating internal scratch space for PPR, and I suspect may need to be marked omp private to prevent the potential race conditions mentioned above.
I must admit I've not experimented with PPR + threading --- is the omp private directive smart enough to protect a struct that contains allocated arrays?

Reply:
@mark-petersen This fix has been included so we're ready for your final review.